File: 이상욱/BOJ/m3w10/BOJ5547.java
package m2w10;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class BOJ5547 {
	
	static int col,row;
	static int[][] map;
	static boolean[][] outside;
	static int[] oddDx = {-1,0,1,1,0,-1};
	static int[] oddDy = {0,-1,0,1,1,1};
	static int[] evenDx = {-1,0,1,1,0,-1};
	static int[] evenDy = {-1,-1,-1,0,1,0};
	static Queue<int[]> q;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		col = Integer.parseInt(st.nextToken());
		row = Integer.parseInt(st.nextToken());
		map = new int[row+2][col+2];
		outside = new boolean[row+2][col+2];
		
		for(int i = 1; i < row+1; i++) {
			st = new StringTokenizer(br.readLine());
			for(int j = 1; j < col+1; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		q = new LinkedList<>();
		
		q.offer(new int[]{0,0});
		outside[0][0] = true;
		
		while(!q.isEmpty()) {
			int[] arr = q.poll();
			int cx = arr[0];
			int cy = arr[1];
			for(int i =0; i < 6; i++) {
				if( cx % 2 == 0) {
					checkOut(evenDx, evenDy, cx, cy, i);
				}else {
					checkOut(oddDx,oddDy, cx, cy, i);
				}
			}
		}
		
		int ans = 0;
		for(int i = 1; i < row+1; i++) {
			for(int j = 1; j < col+1;j++ ) {
				if(map[i][j] == 0 ) {
					continue;
				}
				for(int k = 0; k < 6; k++) {
					int nx = 0;
					int ny = 0;
					if(i%2 == 0) {
						nx = i + evenDx[k];
						ny = j + evenDy[k];
					}else {
						nx = i + oddDx[k];
						ny = j + oddDy[k];
					}
					
					if(outside[nx][ny]) {
						ans++;
					}
				}
			}
		}
		System.out.println(ans);
	}
	private static void checkOut(int[] dx, int[] dy, int cx, int cy, int i) {
		int nx = cx + dx[i];
		int ny = cy + dy[i];
		if(nx >= 0 && nx <= row+1 && ny >= 0 && ny <= col+1) {
			if(!outside[nx][ny]) {
				if(map[nx][ny] == 0) {
					outside[nx][ny] = true;
					q.offer(new int[] {nx,ny});
				}
			}
		}
		
	}
}

File: 이상욱/BOJ/m3w10/BOJ5547.kt
package m3w10

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

var col = 0
var row = 0
lateinit var map: Array<IntArray>
lateinit var outside: Array<BooleanArray>
val oddDx = intArrayOf(-1,0,1,1,0,-1)
val oddDy = intArrayOf(0,-1,0,1,1,1)
val evenDx = intArrayOf(-1,0,1,1,0,-1)
val evenDy = intArrayOf(-1,-1,-1,0,1,0)
lateinit var q : Queue<IntArray>

//참고 사이트 : https://velog.io/@paulus0617/boj5547

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())
    col = st.nextToken().toInt()
    row = st.nextToken().toInt()
    map1 = Array(row + 2) { IntArray(col + 2) }
    outside = Array(row + 2) { BooleanArray(col + 2) }
    for (i in 1 until row + 1) {
        st = StringTokenizer(br.readLine())
        for (j in 1 until col + 1) {
            map1[i][j] = st.nextToken().toInt()
        }
    }
    q = LinkedList()
    q.offer(intArrayOf(0, 0))
    outside[0][0] = true
    while (!q.isEmpty()) {
        val arr = q.poll()
        val cx = arr[0]
        val cy = arr[1]
        for (i in 0 until 6) {
            if (cx % 2 == 0) {
                checkOut(evenDx, evenDy, cx, cy, i)
            } else {
                checkOut(oddDx, oddDy, cx, cy, i)
            }
        }
    }
    var answer = 0
    for (i in 1 until row + 1) {
        for (j in 1 until col + 1) {
            if (map1[i][j] == 0) continue
            for (k in 0 until 6) {
                var nx = 0
                var ny = 0
                if (i % 2 == 0) {
                    nx = i + evenDx[k]
                    ny = j + evenDy[k]
                } else {
                    nx = i + oddDx[k]
                    ny = j + oddDy[k]
                }
                if (outside[nx][ny]) answer++
            }
        }
    }
    println(answer)
}

fun checkOut(dx: IntArray, dy: IntArray, cx: Int, cy: Int, i: Int) {
    val nx = cx + dx[i]
    val ny = cy + dy[i]
    if (nx >= 0 && nx <= row + 1 && ny >= 0 && ny <= col + 1) {
        if (!outside[nx][ny]) {
            if (map1[nx][ny] == 0) {
                outside[nx][ny] = true
                q.offer(intArrayOf(nx, ny))
            }
        }
    }
}